receiver/k8sclusterreceiver/watcher.go

```diff
@@ -95,24 +99,72 @@ func (rw *resourceWatcher) setupInformers(o runtime.Object, informer cache.Share
 }
 
 func (rw *resourceWatcher) onAdd(obj interface{}) {
-	rw.addOrUpdateResource(obj)
+	rw.dataCollector.SyncMetrics(obj)
+
+	// Sync metadata only if there's at least one destination for it to sent.
+	if len(rw.metadataConsumers) == 0 {
+		return
+	}
+
+	newMetadata := rw.dataCollector.SyncMetadata(obj)
+	rw.syncMetadataUpdate(map[string]*collection.KubernetesMetadata{}, newMetadata)
 }
 
 func (rw *resourceWatcher) onDelete(obj interface{}) {
 	rw.dataCollector.RemoveFromMetricsStore(obj)
 }
 
-func (rw *resourceWatcher) onUpdate(_ interface{}, newObj interface{}) {
-	rw.addOrUpdateResource(newObj)
+func (rw *resourceWatcher) onUpdate(oldObj, newObj interface{}) {
+	// Sync metrics from the new object
+	rw.dataCollector.SyncMetrics(newObj)
+
+	// Sync metadata only if there's at least one destination for it to sent.
+	if len(rw.metadataConsumers) == 0 {
+		return
+	}
+
+	oldMetadata := rw.dataCollector.SyncMetadata(oldObj)
+	newMetadata := rw.dataCollector.SyncMetadata(newObj)
+
+	rw.syncMetadataUpdate(oldMetadata, newMetadata)
+}
+
+func (rw *resourceWatcher) setupMetadataExporters(
+	exporters map[configmodels.DataType]map[configmodels.Exporter]component.Exporter,
+	metadataExportersFromConfig []string) error {
+	var out []metadataConsumer
+	metadataExportersSet := utils.StringSliceToMap(metadataExportersFromConfig)
+
+	for pipelineType, exps := range exporters {
+		for cfg, exp := range exps {
+			if !metadataExportersSet[string(cfg.Type())] {
+				continue
+			}
+			kme, ok := exp.(collection.KubernetesMetadataExporter)
+			if !ok {
+				return fmt.Errorf("%v exporter does not implement KubernetesMetadataExporter", cfg.Type())
+			}
+			out = append(out, kme.ConsumeKubernetesMetadata)
+			rw.logger.Info("Configured Kubernetes MetadataExporter",
+				zap.String("pipeline_type", pipelineType.GetString()),
+				zap.String("exporter_type", string(cfg.Type())),
+			)
+		}
+	}
+
+	rw.metadataConsumers = out
+	return nil
 }
 
-// addOrUpdateResource keeps the metric cache updated with latest metric
-// values that will be dispatched in the subsequent interval.
-func (rw *resourceWatcher) addOrUpdateResource(obj interface{}) {
-	rw.dataCollector.SyncMetrics(obj)
+func (rw *resourceWatcher) syncMetadataUpdate(oldMetadata, newMetadata map[string]*collection.KubernetesMetadata) {
+	kubernetesMetadataUpdate := collection.GetKubernetesMetadataUpdate(oldMetadata, newMetadata)
 
-	// Sync metadata only if there's at least one destination for it to sent.
-	if rw.collectMedata {
-		rw.dataCollector.SyncMetadata(obj)
+	if len(kubernetesMetadataUpdate) == 0 {
+		return
+	}
+
+	// TODO: Make this asynchronous
```

Comment:
Between multiple consumers or async to consumers themselves?

Reply:
Between multiple consumers.